bindings/matrix-sdk-crypto-ffi/src/verification.rs

```diff
@@ -1,101 +1,365 @@
+            RustSasState::KeysExchanged { emojis, decimals } => Self::KeysExchanged {
+                emojis: emojis.map(|e| e.indices.map(|i| i as i32).to_vec()),
+                decimals: [decimals.0.into(), decimals.1.into(), decimals.2.into()].to_vec(),
+            },
+            RustSasState::Confirmed => Self::Confirmed,
+            RustSasState::Done { .. } => Self::Done,
+            RustSasState::Cancelled(c) => Self::Cancelled { cancel_info: c.into() },
+        }
+    }
+}
+
+/// Enum representing the different verification flows we support.
+pub struct Verification {
+    pub(crate) inner: InnerVerification,
+    pub(crate) runtime: Handle,
+}
+
+impl Verification {
+    /// Try to represent the `Verification` as an `Sas` verification object,
+    /// returns `None` if the verification is not a `Sas` verification.
+    pub fn as_sas(&self) -> Option<Arc<Sas>> {
+        if let InnerVerification::SasV1(sas) = &self.inner {
+            Some(Sas { inner: sas.to_owned(), runtime: self.runtime.to_owned() }.into())
+        } else {
+            None
+        }
+    }
+
+    /// Try to represent the `Verification` as an `QrCode` verification object,
+    /// returns `None` if the verification is not a `QrCode` verification.
+    pub fn as_qr(&self) -> Option<Arc<QrCode>> {
+        if let InnerVerification::QrV1(qr) = &self.inner {
+            Some(QrCode { inner: qr.to_owned() }.into())
+        } else {
+            None
+        }
+    }
+}
+
 /// The `m.sas.v1` verification flow.
 pub struct Sas {
-    /// The other user that is participating in the verification flow
-    pub other_user_id: String,
-    /// The other user's device that is participating in the verification flow
-    pub other_device_id: String,
-    /// The unique ID of this verification flow, will be a random string for
-    /// to-device events or a event ID for in-room events.
-    pub flow_id: String,
-    /// The room ID where this verification is happening, will be `None` if the
-    /// verification is going through to-device messages
-    pub room_id: Option<String>,
-    /// Did we initiate the verification flow
-    pub we_started: bool,
-    /// Has the non-initiating side accepted the verification flow
-    pub has_been_accepted: bool,
-    /// Can the short auth string be presented
-    pub can_be_presented: bool,
-    /// Does the flow support the emoji representation of the short auth string
-    pub supports_emoji: bool,
-    /// Have we confirmed that the short auth strings match
-    pub have_we_confirmed: bool,
-    /// Has the verification completed successfully
-    pub is_done: bool,
-    /// Has the flow been cancelled
-    pub is_cancelled: bool,
-    /// Information about the cancellation of the flow, will be `None` if the
-    /// flow hasn't been cancelled
-    pub cancel_info: Option<CancelInfo>,
+    pub(crate) inner: InnerSas,
+    pub(crate) runtime: Handle,
+}
+
+impl Sas {
+    /// Get the user id of the other side.
```

Comment:
Could we have a getter for the current state?

Reply:
Added in https://github.com/matrix-org/matrix-rust-sdk/pull/1184/commits/2703a43c1c926cb98e84dac6de874f101b3ea6b7